---
title: "Create Plots for TIER-seq data sets"
author:
- affiliation: AG Annegret Wilde, Institute of Biology III
  name: "Ute Hoffmann"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    fig_width: 15
    fig_height: 8
    theme: united
    toc: yes
    number_sections: true
  pdf_document:
    toc: yes
    number_sections: true
---

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/6_Plot/")
library(tidyverse)
library(DESeq2)
library(edgeR)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(pheatmap)
library(RColorBrewer)
library(vsn)
library(GenomicRanges)
library(Biostrings)
library(rtracklayer)
library(viridis)
library(colorblindr)
library(clusterProfiler)
```

# Read .gffs

```{r}
# 14 int-TSS are duplicates and are also annotated as alt-TSS
features <- rtracklayer::import("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/input/20210217_syne_onlyUnique_withFeat.gff3")
TUs <- rtracklayer::import("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/input/Kopf_4091_TUs_combined.gff3")
# import PSS and TSS identified in TIER-seq analysis
TIERseq_PSS_TSS <- rtracklayer::import("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/output/annoTSS_comparison/gffs/TSS_PSS_rneAnalysis_all.gff")
```

# Read .grps

```{r}
transc_fwd <- read.table("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/output/grp/transcript/AP004311.1_transcript_Ts-0h-1h-WT-0h-1h_fw.grp")
names(transc_fwd) <- c("Ts_0h", "Ts_1h", "WT_0h", "WT_1h")
transc_rev <- read.table("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/output/grp/transcript/AP004311.1_transcript_Ts-0h-1h-WT-0h-1h_rev.grp")
names(transc_rev) <- c("Ts_0h", "Ts_1h", "WT_0h", "WT_1h")
TSS_fwd <- read.table("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/output/grp/TSS/AP004311.1_TSS_Ts-0h-1h-WT-0h-1h_fw.grp")
names(TSS_fwd) <- c("Ts_0h", "Ts_1h", "WT_0h", "WT_1h")
TSS_rev <- read.table("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/output/grp/TSS/AP004311.1_TSS_Ts-0h-1h-WT-0h-1h_rev.grp")
names(TSS_rev) <- c("Ts_0h", "Ts_1h", "WT_0h", "WT_1h")
PSS_fwd <- read.table("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/output/grp/PSS/AP004311.1_PSS_Ts-0h-1h-WT-0h-1h_fw.grp")
names(PSS_fwd) <- c("Ts_0h", "Ts_1h", "WT_0h", "WT_1h")
PSS_rev <- read.table("/data/Dokumente/uni/Doktorarbeit/20201010_TIER-Seq_Data/00_Manuscript/Code_final/4_Analysis_PSS-TSS/output/grp/PSS/AP004311.1_PSS_Ts-0h-1h-WT-0h-1h_rev.grp")
names(PSS_rev) <- c("Ts_0h", "Ts_1h", "WT_0h", "WT_1h")
```

# Plot stuff

```{r}
start=90000
end=91000
subset_data <- transc_fwd[start:end,]
subset_data$nt_pos <- start:end
subset_data_l <- pivot_longer(subset_data, names_to="condition", cols=1:4)
p <- ggplot(subset_data_l, aes(x=nt_pos, y=value, color=condition))+geom_line()+facet_grid(condition~.)
feature_in_region <- subset(features, seqnames(features)=="AP004311.1" & start(features) >= start & end(features) <= end & strand(features)=="+")
p
```

```{r}

colors=c("Ts_0h"="#e69f00ff", "Ts_1h"="#e69f00ff", "WT_0h"="#005a96ff", "WT_1h"="#005a96ff")
linetypes=c("Ts_0h"=2, "Ts_1h"=1, "WT_0h"=2, "WT_1h"=1)

# Function specifically to plot TIER-seq data
plot_coverage <- function(seqname, start, end, color, lty_vector, box_colors, fwd_transc, rev_transc, fwd_TSS, rev_TSS, fwd_PSS, rev_PSS, gff){
  
  tmp_fwd_transc <- log2(fwd_transc[start:end,]+1)+2
  tmp_rev_transc <- log2(rev_transc[start:end,]+1)+2
  
  # plot transcript data
  plot(1,1,xlim=c(start,end),ylim=c(-max(tmp_rev_transc),max(tmp_fwd_transc)),bty="n",ylab="log2(counts)",xlab="",xaxt="n",yaxt="n") # initalize plot
  axis(2,2:20, labels=0:18)
  axis(2,(2:20)*-1, labels=(0:18))
  for(i in 1:length(tmp_fwd_transc)){
    lines(start:end, tmp_fwd_transc[,i], col=colors[names(tmp_fwd_transc)[i]], lty=lty_vector[names(tmp_fwd_transc)[i]]) # plot coverage 
  }
  for(i in 1:length(tmp_rev_transc)){
    lines(start:end, -(tmp_rev_transc[,i]), col=colors[names(tmp_rev_transc)[i]], lty=lty_vector[names(tmp_rev_transc)[i]]) # plot coverage 
  }  
  
  
  # draw horizontal lines to connect boxes on x-axis
  abline(h=c(-2,-1,1,2))
  
  ## gff annotation
  # extract features in region
  feature_in_region_fw <- as.data.frame(subset(gff, seqnames(gff)==seqname & start(gff) >= start & end(gff) <= end & strand(gff)=="+"))
  feature_in_region_rev <- as.data.frame(subset(gff, seqnames(gff)==seqname & start(gff) >= start & end(gff) <= end & strand(gff)=="-")) # some problem here!
 
  # draw rectangles and label them for fw-strand features
  if(nrow(feature_in_region_fw)){
    rect(feature_in_region_fw$start, rep(1.5,nrow(feature_in_region_fw)), feature_in_region_fw$end, rep(0.5,nrow(feature_in_region_fw)), col="gray")
    nam <- feature_in_region_fw$locus_tag
    label_x_values <- (feature_in_region_fw$end - feature_in_region_fw$start)/2+feature_in_region_fw$start
    label_y_values <- rep(1.25, nrow(feature_in_region_fw))
    label_y_values[which((1:nrow(feature_in_region_fw))%%2!=0)]<-(-0.75)
    text(label_x_values, label_y_values, labels=nam, cex=0.5)
  }
  
  # draw rectangles and label them for rev-strand features
  if(nrow(feature_in_region_rev)){
    rect(feature_in_region_rev$start, rep(-0.5,nrow(feature_in_region_rev)), feature_in_region_rev$end, rep(-1.5,nrow(feature_in_region_rev)), col="gray")
    nam <- feature_in_region_rev$locus_tag
    label_x_values <- (feature_in_region_rev$end - feature_in_region_rev$start)/2+feature_in_region_rev$start
    label_y_values <- rep(-1.25, nrow(feature_in_region_rev))
    label_y_values[which((1:nrow(feature_in_region_rev))%%2!=0)]<-(-0.75)
    text(label_x_values, label_y_values, labels=nam, cex=0.5)
  }
  
  ## put legend into upper right corner
  legend("topright", bty="n", legend=names(tmp_fwd_transc), text.col=color)
}

plot_coverage(seqname="AP004311.1", start=89000, end=92000, fwd_transc=transc_fwd, rev_transc=transc_rev, gff=features, color=colors, lty_vector=linetypes)
```

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
